### ISSUE TITLE
Check if a template exists

### ISSUE THREAD

**Issue by Fender123** [casid/jte#52]
Hey,

I'm using a template directory structure containing views for different languages like this:

- views/de/imprint.jte
- views/de/at/imprint.jte

Is there a performant way using TemplateEngine or some other jte code to check whether a template exists?
I.e. for locale de_de I first check if the template views/de/de/imprint.jte exists. If it doesn't exist the next check is views/de/imprint.jte. So it should be sufficient to check whether the classes for each template exists, but I don't want to reimplement jte class name generation. And in not precompiled mode the classes are generated on the fly, so the check should look different.

I hope you get the point and tell me whether this is something you could do in jte or something you don't want to have in jte codebase :-)

**Comment by casid** [comment]
Hey @Fender123, not right now without causing and catching unneccary exceptions by the `TemplateLoader`...

But I think we can add a method for that to `TemplateEngine`, like `public boolean hasTemplate(String name)`.

**Comment by casid** [comment]
@Fender123 I've added a first draft of the `hasTemplate` method: https://github.com/casid/jte/commit/cc13ba8ba1463d25dd66c8b577ed0d907177a0fe

There is now a `TemplateNotFoundException` that derives from `TemplateException`. It is only thrown if a template cannot be resolved. `hasTemplate` uses this internally.

With the new Exception, you could now also do this:

```java
try {
  render(specialLocale);
} catch(TemplateNotFoundException e) {
  render(fallbackLocale);
}
```java

**Comment by Fender123** [comment]
Thanks, this looks promising

**Comment by casid** [comment]
@Fender123 I plan to release jte 1.7.0 soon, this will be part of that release!

**Comment by casid** [comment]
Released with 1.7.0 👍

**Comment by Fender123** [comment]
hey, sorry to bother you, but after implementing `hasTemplate` in our codebase, I saw that `DirectoryCodeResolver` throws an unhandled  java.nio.file.NoSuchFileException when calling `hasTemplate` with a not existing template. It is working fine for precompiled template engine. 
I think this should be fixed in `hasTemplate` and not in the code using that function.

**Comment by casid** [comment]
Oh, sorry about that. Classic case of unit test => green, integration test => not so green.

I pushed a fix for `DirectoryCodeResolver` 👍 

Thanks for reporting!

**Comment by casid** [comment]
Hey @Fender123, this fix is finally released with 1.8.0 👍
